cmd/oras/upload.go

```diff
@@ -0,0 +1,156 @@
+** This command is in preview and under development. **
+
+Example - uplad artifacts
+  oras upload localhost:5000/owner/this:v1 localhost:6000/owner/that:v1
+`,
+		Args: cobra.ExactArgs(1),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return option.Parse(&opts)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runupload(opts, args[0])
+		},
+	}
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runupload(opts uploadOptions, destination string) error {
+	ctx, _ := opts.SetLoggerLevel()
+
+	var uniqueImages []registry.Reference
+	err := filepath.Walk(directory,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ociPath, layout := strings.CutSuffix(path, "/oci-layout")
+			if layout == true {
+				ociPath, _ = strings.CutPrefix(ociPath, directory+"/")
+				lastIdx := strings.LastIndex(ociPath, "/")
+				reference := registry.Reference{
+					Repository: ociPath[:lastIdx],
+					Reference:  ociPath[lastIdx+1:],
+				}
+				uniqueImages = append(uniqueImages, reference)
+			}
+			return nil
+		})
+	if err != nil {
+		return fmt.Errorf("problem traversing directory %s: %v", directory, err)
+	}
+
+	// Prepare upload options
+	var src option.ReadOnlyGraphTagFinderTarget
+	committed := &sync.Map{}
+	extendedCopyOptions := oras.DefaultExtendedCopyOptions
+	extendedCopyOptions.Concurrency = concurrency
+	extendedCopyOptions.PreCopy = display.StatusPrinter("uploading", opts.Verbose)
+	extendedCopyOptions.PostCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		if err := display.PrintSuccessorStatus(ctx, desc, "Skipped", src, committed, opts.Verbose); err != nil {
+			return err
+		}
+		return display.PrintStatus(desc, "uploaded ", opts.Verbose)
+	}
+	extendedCopyOptions.OnCopySkipped = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		return display.PrintStatus(desc, "Exists ", opts.Verbose)
+	}
+	uploadOptions := oras.CopyOptions{
+		CopyGraphOptions: extendedCopyOptions.CopyGraphOptions,
+	}
+
+	for _, fileReference := range uniqueImages {
+		// Prepare source
+		var from option.Target
+		from.Type = option.TargetTypeOCILayout
+		from.Path = path.Join(directory, fileReference.Repository, fileReference.Reference)
+		from.Reference = fileReference.Reference
+		from.RawReference = from.Path
+		src, err = oci.NewFromFS(ctx, os.DirFS(from.Path))
+		if err != nil {
+			return err
+		}
+
+		// Prepare destination
+		fileReference.Registry = destination
+		to := option.Target{
+			Path:         path.Join(destination, fileReference.Repository),
+			Reference:    fileReference.Reference,
+			RawReference: fileReference.String(),
+			Remote:       opts.To,
+			Type:         option.TargetTypeRemote,
+		}
+
+		to.Reference = ""
+		dst, err := to.NewTarget(opts.Common)
```

Comment:
Since the target cannot be a oci layout, you can use opts.To.NewRepository directly